### PR TITLE
feat: multi-threading

### PR DESCRIPTION
**Pull request by AirQuick**
This pull request adds a new attribute, `@threads`, to `x:description` and `x:scenario`. For example,

```xml
<x:description threads="2">
    <x:scenario label="scenario #1">...</x:scenario>
    <x:scenario label="scenario #2">...</x:scenario>
</x:description>
```

```xml
<x:description>
    <x:scenario threads="2">
        <x:scenario label="scenario #1">...</x:scenario>
        <x:scenario label="scenario #2">...</x:scenario>
    </x:scenario>
</x:description>
```

In both cases, `scenario #1` and `scenario #2` will be executed with 2 threads. So they will run at the same time.

`@threads` works only on child scenarios. For example,

```xml
<x:description threads="2">
    <x:scenario label="scenario #1">
        <x:scenario label="scenario #1.1">...</x:scenario>
        <x:scenario label="scenario #1.2">...</x:scenario>
    </x:scenario>
    <x:scenario label="scenario #2">...</x:scenario>
</x:description>
```

`scenario #1.1` and `scenario #1.2` are not multi-threaded. They will run sequentially while `scenario #1` and `scenario #2` will run in parallel.

The value of `@threads` is an XPath expression evaluated at run time. Usually it will just a positive integer atomic value, but you can use other expressions.

Two special values are available for your convenience:

- `#child-scenario-count` means the number of child scenarios:

  ```xml
  <x:scenario threads="#child-scenario-count">
      <x:scenario>...</x:scenario>
      <x:scenario>...</x:scenario>
  </x:scenario>
  ```

  is equal to

  ```xml
  <x:scenario threads="2">
      <x:scenario>...</x:scenario>
      <x:scenario>...</x:scenario>
  </x:scenario>
  ```

- `#logical-processor-count` means the number of logical processors:

  If you have 4 logical processors, `threads="#logical-processor-count"` is equal to `threads="4"`.

### Technical details

This feature entirely depends on [`@saxon:threads`](https://www.saxonica.com/documentation10/index.html#!extensions/attributes/threads).

### Requirements

- Requires Saxon-EE. Ignored on the other processors.
- Only for XSLT and Schematron. Ignored on XQuery.

### Tests

Testing this feature is a bit complicated by its nature.

There are 8 test files (`test/threads_*.xspec`) and their support files (`test/threads/*.*`).

- `test/threads_*_ignored*.xspec` files verify that `@threads` is ignored as expected.
- The other `test/threads_*.xspec` files, together with `test/threads/*.*`, verify that `@threads` takes effect as expected:
  - `test/threads/sleeper*.*` files provide a basic scenario (the `Sleep for a period` scenario) which sleeps for milliseconds specified by `@ms` in `sleeper.xml` (= 5 seconds).
  - `test/threads/test*.*` files combine 5 copies of the basic scenario. So, without `@threads`, the combined scenarios will take `5 seconds * 5 iterations` = 25 seconds (+ some overhead).
  - `@measure-time` is enabled so that the elapsed times are recorded.
  - `test/threads/check-xml-report.xsl` works in place of `src/reporter/format-xspec-report.xsl` and verifies the elapsed times recorded in the test result report XML. For example:
    - If `@threads` is 2, the combined 5 scenarios will take `ceiling(5 scenarios / 2 threads)` = 3 iterations. The elapsed time should be `5 seconds * 3 iterations` = 15 seconds (+ some overhead).
    - If `@threads` is 4, the combined 5 scenarios will take `ceiling(5 scenarios / 4 threads)` = 2 iterations. The elapsed time should be `5 seconds * 2 iterations` = 10 seconds (+ some overhead).
    - If `@threads` is 5 or larger, the combined 5 scenarios will take `ceiling(5 scenarios / 5+ threads)` = 1 iteration. The elapsed time should be `5 seconds * 1 iteration` = 5 seconds (+ some overhead).

Here's how I ran the test with Saxon-EE:

<details>
    <summary>Windows + 8 logical processors + Saxon-EE 9.9.1.8</summary>

```console
C:\xspec>test\ci\install-deps.cmd
Setting up saxon-9-9
...
BUILD SUCCESSFUL
Total time: 11 seconds

C:\xspec>set SAXON_JAR=c:\SaxonEE9-9-1-8J\saxon9ee.jar

C:\xspec>test\run-xspec-tests-ant.cmd -silent -Dthread.count=1 -Dxspecfiles.dir.url.query="select=threads_*.xspec"
Buildfile: C:\xspec\test\ant\build.xml
Getting processor capabilities
...
Testing threads_description_ignored_no-scenario.xspec [q]


total: 0

Testing threads_description_ignored_no-scenario.xspec [s]


total: 0

Testing threads_description_ignored_no-scenario.xspec [t]


total: 0

Testing threads_description_ignored_not-supported.xspec [q]


passed: 2 / pending: 0 / failed: 0 / total: 2

Testing threads_description_ignored_not-supported.xspec [s]


passed: 2 / pending: 0 / failed: 0 / total: 2

Testing threads_description_ignored_not-supported.xspec [t]


passed: 2 / pending: 0 / failed: 0 / total: 2

Testing threads_description_ignored_one-scenario.xspec [q]


passed: 1 / pending: 0 / failed: 0 / total: 1

Testing threads_description_ignored_one-scenario.xspec [s]


passed: 1 / pending: 0 / failed: 0 / total: 1

Testing threads_description_ignored_one-scenario.xspec [t]


passed: 1 / pending: 0 / failed: 0 / total: 1

Testing threads_description_schematron.xspec [s]


5 scenarios took PT6.125038S. Expected PT5S ~ PT7S with 5 thread(s).

passed: 5 / pending: 0 / failed: 0 / total: 5

Testing threads_description_stylesheet.xspec [t]


5 scenarios took PT5.406467S. Expected PT5S ~ PT7S with 5 thread(s).

passed: 5 / pending: 0 / failed: 0 / total: 5

Testing threads_scenario_ignored.xspec [q]


passed: 5 / pending: 0 / failed: 0 / total: 5

Testing threads_scenario_ignored.xspec [s]


passed: 5 / pending: 0 / failed: 0 / total: 5

Testing threads_scenario_ignored.xspec [t]


passed: 5 / pending: 0 / failed: 0 / total: 5

Testing threads_scenario_schematron.xspec [s]


'#child-scenario-count' scenario containing 5 scenarios took PT6.031747S. Expected PT5S ~ PT7S with 5 thread(s).

'#logical-processor-count' scenario containing 5 scenarios took PT5.140067S. Expected PT5S ~ PT7S with 8 thread(s).

'XPath expression using local namespace declaration' scenario containing 5 scenarios took PT10.016141S. Expected PT10S ~ PT12S with 4 thread(s).

passed: 15 / pending: 0 / failed: 0 / total: 15

Testing threads_scenario_stylesheet.xspec [t]


'#child-scenario-count' scenario containing 5 scenarios took PT5.296867S. Expected PT5S ~ PT7S with 5 thread(s).

'#logical-processor-count' scenario containing 5 scenarios took PT5.031668S. Expected PT5S ~ PT7S with 8 thread(s).

'XPath expression using local namespace declaration' scenario containing 5 scenarios took PT10.015545S. Expected PT10S ~ PT12S with 4 thread(s).

passed: 15 / pending: 0 / failed: 0 / total: 15


C:\xspec>
```

</details>

<details>
   <summary>Linux + 2 logical processors + Saxon-EE 10.5</summary>

```console
$ export XSPEC_TEST_ENV=saxon-10
$ source test/ci/install-deps.sh 
Setting up saxon-10
...
BUILD SUCCESSFUL
Total time: 22 seconds
$ export SAXON_JAR=/tmp/SaxonEE10-5J/saxon-ee-10.5.jar
$ test/run-xspec-tests-ant.sh -silent -Dthread.count=1 -Dxspecfiles.dir.url.query='select=threads_*.xspec'
Buildfile: /tmp/xspec/test/ant/build.xml
Getting processor capabilities
...
Testing threads_description_ignored_no-scenario.xspec [q]
 
total: 0
Testing threads_description_ignored_no-scenario.xspec [s]
 
total: 0
Testing threads_description_ignored_no-scenario.xspec [t]
 
total: 0
Testing threads_description_ignored_not-supported.xspec [q]
 
passed: 2 / pending: 0 / failed: 0 / total: 2
Testing threads_description_ignored_not-supported.xspec [s]
 
passed: 2 / pending: 0 / failed: 0 / total: 2
Testing threads_description_ignored_not-supported.xspec [t]
 
passed: 2 / pending: 0 / failed: 0 / total: 2
Testing threads_description_ignored_one-scenario.xspec [q]
 
passed: 1 / pending: 0 / failed: 0 / total: 1
Testing threads_description_ignored_one-scenario.xspec [s]
 
passed: 1 / pending: 0 / failed: 0 / total: 1
Testing threads_description_ignored_one-scenario.xspec [t]
 
passed: 1 / pending: 0 / failed: 0 / total: 1
Testing threads_description_schematron.xspec [s]
 
5 scenarios took PT5.196S. Expected PT5S ~ PT7S with 5 thread(s).
passed: 5 / pending: 0 / failed: 0 / total: 5
Testing threads_description_stylesheet.xspec [t]
 
5 scenarios took PT5.161S. Expected PT5S ~ PT7S with 5 thread(s).
passed: 5 / pending: 0 / failed: 0 / total: 5
Testing threads_scenario_ignored.xspec [q]
 
passed: 5 / pending: 0 / failed: 0 / total: 5
Testing threads_scenario_ignored.xspec [s]
 
passed: 5 / pending: 0 / failed: 0 / total: 5
Testing threads_scenario_ignored.xspec [t]
 
passed: 5 / pending: 0 / failed: 0 / total: 5
Testing threads_scenario_schematron.xspec [s]
 
'#child-scenario-count' scenario containing 5 scenarios took PT5.329S. Expected PT5S ~ PT7S with 5 thread(s).
'#logical-processor-count' scenario containing 5 scenarios took PT15.042S. Expected PT15S ~ PT17S with 2 thread(s).
'XPath expression using local namespace declaration' scenario containing 5 scenarios took PT10.019S. Expected PT10S ~ PT12S with 4 thread(s).
passed: 15 / pending: 0 / failed: 0 / total: 15
Testing threads_scenario_stylesheet.xspec [t]
 
'#child-scenario-count' scenario containing 5 scenarios took PT5.174S. Expected PT5S ~ PT7S with 5 thread(s).
'#logical-processor-count' scenario containing 5 scenarios took PT15.025S. Expected PT15S ~ PT17S with 2 thread(s).
'XPath expression using local namespace declaration' scenario containing 5 scenarios took PT10.025S. Expected PT10S ~ PT12S with 4 thread(s).
passed: 15 / pending: 0 / failed: 0 / total: 15
$ 
```

</details>

Saxon-HE is tested on our usual CI systems.

### CODE REVIEW

**Comment by galtm** [comment]
@AirQuick , Looks interesting! I should be able to review this over the weekend.

File: test/ant/worker/generate.xsl [hunk 1]
```diff
@@ -159,6 +159,13 @@
 						<xsl:text>Requires XSLT processor to support Java reflexive extension functions</xsl:text>
 					</xsl:when>
 
+					<xsl:when test="
+							($test-type eq 't')
```
Comment (galtm) [inline test/ant/worker/generate.xsl:163]: I'm not very familiar with this file, but I wondered why this logic doesn't apply to both `'t'` and `'s'` types.

**Review by galtm — COMMENTED** [review]
I didn't see a test for `threads="#logical-processor-count"` on `x:description`. I think that is OK because of how similar the treatment of `@threads` is for `x:description` vs. `x:scenario` in the compiler, but I'm mentioning this just in case you meant to cover `<x:description threads="#logical-processor-count" ...>` explicitly.

Regarding the ability to make `@threads` hold an XPath expression: It would be good to balance this flexibility with more error checking, to avoid error messages that point to the compiled stylesheet in cases like `threads=""` or XPath expressions that are nonempty but don't evaluate to a positive integer. Maybe `threads=""` should use single-threaded behavior with no error message.

File: test/ant/worker/generate.xsl (hunk 1 above)
Comment (AirQuick) [reply test/ant/worker/generate.xsl]: You're right. Fixed in ca51c236918972ecb184e33b60a178a475272f13.

**Comment by AirQuick** [comment]
Sorry for not responding sooner.

> I didn't see a test for `threads="#logical-processor-count"` on `x:description`. I think that is OK because of how similar the treatment of `@threads` is for `x:description` vs. `x:scenario` in the compiler, but I'm mentioning this just in case you meant to cover `<x:description threads="#logical-processor-count" ...>` explicitly.

`<x:description threads="#logical-processor-count" ...>` is now tested: 21a53c24becf27d3a9aee01691e7183b0d714fb2

> Regarding the ability to make `@threads` hold an XPath expression: It would be good to balance this flexibility with more error checking, to avoid error messages that point to the compiled stylesheet in cases like `threads=""` or XPath expressions that are nonempty but don't evaluate to a positive integer. Maybe `threads=""` should use single-threaded behavior with no error message.

- 1881ca607126ee1b2d78e73ac1f0312fbfd089af changed the definition of `xpath` in `src/schemas/xspec.rnc`. Zero-length value is now invalid. What's your take on this?

- 4b56ffc69ba792f083a8fe15792c96a00f2c807b added some error checks of `@threads` expression at run time. Now the expression must be an instance of `xs:integer` and >=1.
  [`({$threads-attr})[if (. castable as xs:positiveInteger) then true() else error()]`](https://github.com/xspec/xspec/pull/1497/commits/4b56ffc69ba792f083a8fe15792c96a00f2c807b#diff-86b980547f8ee88a45e11f31f48cb579b4b178fd5c5b5b6bf66555a2f3472284R80) may look weird in that it is a predicate and it is not `if (({$threads-attr}) ge 1) then ... else ...`. But if I write the condition in other ways, Saxon detects an error at static analysis and fails before invoking `error()`.


Here's how I ran the test with Saxon-EE:

<details>
    <summary>Windows + 8 logical processors + Saxon-EE 9.9.1.8</summary>

```console
C:\xspec>test\ci\install-deps.cmd
Setting up saxon-9-9
...
BUILD SUCCESSFUL
Total time: 40 seconds

C:\xspec>set SAXON_JAR=c:\SaxonEE9-9-1-8J\saxon9ee.jar

C:\xspec>test\run-xspec-tests-ant.cmd -silent -Dthread.count=1 -Dxspecfiles.dir.url.query="select=threads_*.xspec"
Buildfile: C:\xspec\test\ant\build.xml
Getting processor capabilities
...
Testing threads_description_ignored_no-scenario.xspec [q]


total: 0

Testing threads_description_ignored_no-scenario.xspec [s]


total: 0

Testing threads_description_ignored_no-scenario.xspec [t]


total: 0

Testing threads_description_ignored_not-supported.xspec [q]


passed: 2 / pending: 0 / failed: 0 / total: 2

Testing threads_description_ignored_not-supported.xspec [s]


passed: 2 / pending: 0 / failed: 0 / total: 2

Testing threads_description_ignored_not-supported.xspec [t]


passed: 2 / pending: 0 / failed: 0 / total: 2

Testing threads_description_ignored_one-scenario.xspec [q]


passed: 1 / pending: 0 / failed: 0 / total: 1

Testing threads_description_ignored_one-scenario.xspec [s]


passed: 1 / pending: 0 / failed: 0 / total: 1

Testing threads_description_ignored_one-scenario.xspec [t]


passed: 1 / pending: 0 / failed: 0 / total: 1

Testing threads_description_schematron.xspec [s]


5 scenarios took PT6.06S. Expected PT5S to PT7S with 5 thread(s).

passed: 5 / pending: 0 / failed: 0 / total: 5

Testing threads_description_stylesheet.xspec [t]


5 scenarios took PT5.215S. Expected PT5S to PT7S with 8 thread(s).

passed: 5 / pending: 0 / failed: 0 / total: 5

Testing threads_scenario_ignored.xspec [q]


passed: 5 / pending: 0 / failed: 0 / total: 5

Testing threads_scenario_ignored.xspec [s]


passed: 5 / pending: 0 / failed: 0 / total: 5

Testing threads_scenario_ignored.xspec [t]


passed: 5 / pending: 0 / failed: 0 / total: 5

Testing threads_scenario_schematron.xspec [s]


'#child-scenario-count' scenario containing 5 scenarios took PT6.13S. Expected PT5S to PT7S with 5 thread(s).

'#logical-processor-count' scenario containing 5 scenarios took PT5.186S. Expected PT5S to PT7S with 8 thread(s).

'XPath expression using local namespace declaration' scenario containing 5 scenarios took PT10.043S. Expected PT10S to PT12S with 4 thread(s).

passed: 15 / pending: 0 / failed: 0 / total: 15

Testing threads_scenario_stylesheet.xspec [t]


'#child-scenario-count' scenario containing 5 scenarios took PT6.2S. Expected PT5S to PT7S with 5 thread(s).

'#logical-processor-count' scenario containing 5 scenarios took PT5.03S. Expected PT5S to PT7S with 8 thread(s).

'XPath expression using local namespace declaration' scenario containing 5 scenarios took PT10.04S. Expected PT10S to PT12S with 4 thread(s).

passed: 15 / pending: 0 / failed: 0 / total: 15


C:\xspec>test\run-bats --filter "@threads"
3 test case(s) (Filter: "@threads")
=== START TEST CASES ================================================
CASE #1: @threads is zero
...:run @  0:54:47.34: ..\bin\xspec.bat threads\dynamic-error\description_zero.xspec
...Verified: Return value: 2
...Verified: Line 7
...PASS
CASE #2: @threads contains more than one item
...:run @  0:54:54.18: ..\bin\xspec.bat threads\dynamic-error\scenario_multiple.xspec
...Verified: Return value: 2
...Verified: Line 8
...PASS
CASE #3: @threads is a string
...:run @  0:55:00.99: ..\bin\xspec.bat threads\dynamic-error\scenario_string.xspec
...Verified: Return value: 2
...Verified: Line 8
...PASS
=== END TEST CASES ==================================================
EXIT_CODE=0

C:\xspec>
```
</details>

<details>
    <summary>Linux + 2 logical processors + Saxon-EE 10.5</summary>

```console
$ export XSPEC_TEST_ENV=saxon-10
$ source test/ci/install-deps.sh
Setting up saxon-10
...
BUILD SUCCESSFUL
Total time: 47 seconds
$ export SAXON_JAR=/tmp/SaxonEE10-5J/saxon-ee-10.5.jar
$ test/run-xspec-tests-ant.sh -silent -Dthread.count=1 -Dxspecfiles.dir.url.query='select=threads_*.xspec'
Buildfile: /tmp/xspec/test/ant/build.xml
Getting processor capabilities
...
Testing threads_description_ignored_no-scenario.xspec [q]
 
total: 0
Testing threads_description_ignored_no-scenario.xspec [s]
 
total: 0
Testing threads_description_ignored_no-scenario.xspec [t]
 
total: 0
Testing threads_description_ignored_not-supported.xspec [q]
 
passed: 2 / pending: 0 / failed: 0 / total: 2
Testing threads_description_ignored_not-supported.xspec [s]
 
passed: 2 / pending: 0 / failed: 0 / total: 2
Testing threads_description_ignored_not-supported.xspec [t]
 
passed: 2 / pending: 0 / failed: 0 / total: 2
Testing threads_description_ignored_one-scenario.xspec [q]
 
passed: 1 / pending: 0 / failed: 0 / total: 1
Testing threads_description_ignored_one-scenario.xspec [s]
 
passed: 1 / pending: 0 / failed: 0 / total: 1
Testing threads_description_ignored_one-scenario.xspec [t]
 
passed: 1 / pending: 0 / failed: 0 / total: 1
Testing threads_description_schematron.xspec [s]
 
5 scenarios took PT5.22S. Expected PT5S to PT7S with 5 thread(s).
passed: 5 / pending: 0 / failed: 0 / total: 5
Testing threads_description_stylesheet.xspec [t]
 
5 scenarios took PT15.192S. Expected PT15S to PT17S with 2 thread(s).
passed: 5 / pending: 0 / failed: 0 / total: 5
Testing threads_scenario_ignored.xspec [q]
 
passed: 5 / pending: 0 / failed: 0 / total: 5
Testing threads_scenario_ignored.xspec [s]
 
passed: 5 / pending: 0 / failed: 0 / total: 5
Testing threads_scenario_ignored.xspec [t]
 
passed: 5 / pending: 0 / failed: 0 / total: 5
Testing threads_scenario_schematron.xspec [s]
 
'#child-scenario-count' scenario containing 5 scenarios took PT5.408S. Expected PT5S to PT7S with 5 thread(s).
'#logical-processor-count' scenario containing 5 scenarios took PT15.027S. Expected PT15S to PT17S with 2 thread(s).
'XPath expression using local namespace declaration' scenario containing 5 scenarios took PT10.018S. Expected PT10S to PT12S with 4 thread(s).
passed: 15 / pending: 0 / failed: 0 / total: 15
Testing threads_scenario_stylesheet.xspec [t]
 
'#child-scenario-count' scenario containing 5 scenarios took PT5.158S. Expected PT5S to PT7S with 5 thread(s).
'#logical-processor-count' scenario containing 5 scenarios took PT15.02S. Expected PT15S to PT17S with 2 thread(s).
'XPath expression using local namespace declaration' scenario containing 5 scenarios took PT10.021S. Expected PT10S to PT12S with 4 thread(s).
passed: 15 / pending: 0 / failed: 0 / total: 15
$ test/run-bats.sh --filter "@threads"
 ✓ @threads is zero 
 ✓ @threads contains more than one item 
 ✓ @threads is a string 

3 tests, 0 failures

$ 
```
</details>

**Comment by galtm** [comment]
>Zero-length value is now invalid. What's your take on this?

Sounds good. I noticed that your change in `xpath` affects not only `@threads` but also `@select`, `@test`, and `@location`. I think that's fine because I can't think of any reason to have zero-length values for any of those attributes.

> * if I write the condition in other ways, Saxon detects an error at static analysis

Thanks for explaining why you did it this way.

**Review by galtm — APPROVED** [review]
Looks great. Thanks for the extra error-checking and testing you added recently.

**Comment by AirQuick** [comment]
@galtm Many thanks for the review! I'm merging this.

**Pull request merged**